### PR TITLE
Added random matrices(Gaussian Ensembles only)

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs

[1] https://github.com/sympy/sympy/issues/17039

#### Brief description of what is fixed or changed
I have added framework random matrices.

#### Other comments
This is a work in progress. The first commit depicts how the framework for random matrices will look like. More features will added in the upcoming updates.

#### Release Notes

* stats
  * Random matrices have been added to `sympy.stats`

### CODE REVIEW

**Comment by czgdp1807** [comment]
ping @Upabjojr @sidhantnagpal 
Please keep tracking, reviewing this PR. Thanks.

File: sympy/stats/random_matrix_models.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        n = self.dimension
+        return 2**(S(n)/2) * pi**(S(n**2)/2)
+
+    def density(self):
+        n, ZGUE = self.dimension, self.normalization_constant
+        h_pspace = RandomMatrixPSpace('P', model=self)
+        H = RandomMatrixSymbol('H', n, n, pspace=h_pspace)
+        return Lambda(H, exp(S(n)/2 * Trace(H**2))/ZGUE)
+
+GUE = GaussianUnitaryEnsemble
```
Comment (sidhantnagpal) [inline sympy/stats/random_matrix_models.py]: Can this be left to the discretion of the user (like imported `as`)? Its usage in the codebase looks rather cryptic.

**Review by Upabjojr — COMMENTED** [review]
Did you also test expressions of random matrices?

**Comment by czgdp1807** [comment]
> Did you also test expressions of random matrices?

Thanks for this case. I will add it in tests. Will update the PR with more features.

**Comment by czgdp1807** [comment]
I have made some minor architectural changes to `GaussianEnsemble`. I am exploring the literature to see how random matrices are added and multiplied and other expressions as I think that finding the density of sum of two random matrix symbols isn't as simple as finding those of random symbols. I will make an update on this. Thanks.

**Comment by Upabjojr** [comment]
> I am exploring the literature to see how random matrices are added and multiplied and other expressions as I think that finding the density of sum of two random matrix symbols isn't as simple as finding those of random symbols.

There is still the possibility of leaving the expressions unevaluated or raising `NotImplementedError`. I just want this to be tested. There is no need to implement this now, just make sure that the code doesn't return wrong results.

**Comment by Upabjojr** [comment]
An exception warning the user that the code is not complete is much better than a wrong result.

**Comment by czgdp1807** [comment]
I will see if I can implement them in this PR. If I found that it will take longer than I will just raise `NotImplementedError` for expressions containing `RandomMatrixSymbols`. In fact, testing for expressions of random matrices helped in understanding the importance of it, so all depends on the efforts required which will decide whether a new PR has to be made after this one.

**Comment by czgdp1807** [comment]
I studied about expression of random matrices. Here are my comments,

**General Observation**

I got to know that there isn't any general technique to find the density of an expression containing more than 1 random matrix. In fact, for example, if we take a matrix from Gaussain Unitary Ensemble(GUE) and raise to a power more than 2 then it's not necessary that it belongs to the GUE(reference - http://web.mit.edu/18.06-esg/18.034/03pract.pdf).

**Some special techniques**

1. If we sum two or more matrices from GUE then it is guaranteed that the result will belong to GUE(concluded from http://web.mit.edu/18.06-esg/18.034/03pract.pdf point - D, page - 2).  So we can find the density with ease when it comes to only sums of GUE.

2. Similarly, if we sum two or more matrices coming from both GUE and Gaussian Orthogonal Ensemble(GOE), because all real symmetric matrices are Hermitian too. 

In short, for sum the result can be determined. 

I suggest that we should raise `NotImplementedError` for expressions which are not `RandomMatrixSymbol` for now. May be some time later, I will make the changes and add the special technique for sums of GUE and GOE, if until then no general solution is not found. Trying out Wolfram will also be a good idea and may help to find the solution.

I will update the PR with other matrix ensembles. Thanks.

File: sympy/stats/random_matrix.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    for handling the API calls for random matrices.
+    """
+    def __new__(cls, sym, model=None):
+        sym = _symbol_converter(sym)
+        return Basic.__new__(cls, sym, model)
+
+    model = property(lambda self: self.args[1])
+
+    def compute_density(self, expr, *args):
+        rms = rms = expr.atoms(RandomMatrixSymbol)
```
Comment (sidhantnagpal) [inline sympy/stats/random_matrix.py]: Duplicated `rms = `.

File: sympy/stats/tests/test_random_matrix.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    _H = MatrixSymbol('_H', 3, 3)
+    G = GOE('O', 3)
+    assert density(G)(H) == exp(-3*Trace(H**2)/4)/Integral(exp(-3*Trace(_H**2)/4), _H)
+    i, j = (Dummy('i', integer=True, positive=True),
+            Dummy('j', integer=True, positive=True))
+    l = IndexedBase('l')
+    assert joint_eigen_distribution(G).dummy_eq(
+    Lambda((l[1], l[2], l[3]),
+    9*sqrt(2)*exp(-3*l[1]**2/2 - 3*l[2]**2/2 - 3*l[3]**2/2)*
+    Product(Abs(l[i] - l[j]), (j, i + 1, 3), (i, 1, 2))/(32*pi)))
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_random_matrix.py]: Minor comment: This (and other places) can be made to match visual indent.

Comment (czgdp1807) [reply sympy/stats/tests/test_random_matrix.py]: Did you mean to align the content of L37, L38, L39 with `(` in L36?

**Comment by czgdp1807** [comment]
This PR is complete, I will add the matrices with random elements as their entries in a new PR rather than overloading this and unmanageable.
